# Part 3 Maji Ndogo project: From analysis to action:

## 
The goals of this part of the project are

1. Integrating the auditors report data into our database: an independent auditor was approached to authenticate the water sources, pollution levels and the residents comments on the surveyors conduct

 2.Generating an ERD from the database
 
3.Linking records: joining our employee data to the auditor report

4. Seeking the truth: bulding a complex query to investigate employee corruption:

In [1]:
%load_ext sql

In [2]:
%sql mysql+pymysql://root:L0xbysmrben%23@localhost:3306/md_water_services

## Importing the Auditor's Report

Now that we have a proper representation of the relationships in our database, we can proceed to import the data from the auditor's report which is in a .csv format. To do this, we need to follow the steps below

1. Create an empty auditor_report table in the md_water_services database. To do this, we run the following in MySQL Workbench:

    ```sql
    DROP TABLE IF EXISTS `auditor_report`;

    CREATE TABLE `auditor_report` (
        `location_id` VARCHAR(32),
        `type_of_water_source` VARCHAR(64),
        `true_water_source_score` INT DEFAULT NULL,
        `statements` VARCHAR(255)
    );
2. Import the data sent by the auditor in .csv format on MySQL Workbench. Remember to use an existing table since we've already created an empty table from the first step.from the first step.

# Auditor's Report Integration

After we've imported the auditor's report into SQL database, at first glance, we can see that it has 1620 records alluding to all revisited sites by the auditor. The report also has the following attributes:
- `location_id` for the revisited locations.
- `type_of_water_source` that was visited by the auditor.
- `true_water_source_score` assigned by the auditor as a measure of water quality.
- `statements` captured while the auditor investigated each site by speaking to locals at the sites.

Based on the auditors report, we can perform a comparative analysis against the surveyors rater_quality` tables.55)
);

## Questions to Answer

Now that we have all our data in one place, let's try and answer the following questions:

1. Is there a difference in scores between the auditor's report and the data provided by the surveyors?
2. If a difference exists, is there a pattern we can identify?AR(255)
    );

## Investigating Differences in Auditor & Surveyor Scores

To investigate if there are differences between auditor and surveyor water quality scores, we will have to perform some joins. Consider the following:
- The `auditor_report` table in our database has a `location_id` attribute but the `water_quality` table only has `record_id` attribute which considers with a similar attribute in the `visits` table.
- The `visits` table has both `location_id` and `record_id` attributes hence it is the perfect table to make the join and link the `auditor_report` and `water_quality` tables.

In [3]:
%sql SHOW TABLES;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
13 rows affected.


[('auditor_report',),
 ('combined_anallysis_table',),
 ('data_dictionary',),
 ('employee',),
 ('global_water_access',),
 ('incorrect_records',),
 ('incorrect_tables',),
 ('location',),
 ('project_progress',),
 ('visits',),
 ('water_quality',),
 ('water_source',),
 ('well_pollution',)]

to verify the authenticity of the database, an audit was commisioned and the finding recorded in the auditor_report table. The auditors report records verified type of water source with its verified water source score as well as its location id. the auditor also included a column on remarks of the residents about the conduct of the surveyors.

In [4]:
%%sql 
SELECT * FROM auditor_report
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[('SoRu34980', 'well', 1, "Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach."),
 ('AkRu08112', 'well', 3, "Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation."),
 ('AkLu02044', 'river', 0, "Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude."),
 ('AkHa00421', 'well', 3, "Villagers were moved by the official's visit, praising their hard work, humility, and the profound sense of connection they fostered."),
 ('SoRu35221', 'river', 0, "A photographer's lens captures the queue, though his own struggle for water is a hidden part of the story."),
 ('HaAm16170', 'well', 1, 'With an open heart, the official created an atmosphere of unity and familial camaraderie among the villagers.'),
 ('AkRu04812', 'well', 3, "The official's presence left an indelible mark, reflecting their humility, dedication, and the genuine connections they nurtured."),
 ('AkRu08304', 'well', 3, "The official's interactions resonated deeply with the villagers, leaving a lasting impression of respect and camaraderie."),
 ('AkRu05107', 'well', 2, "Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation."),
 ('AkRu05215', 'well', 3, "Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern.")]

We confirm that the auditor's report contains a total of **1620** records shown by the output from running the cell above. These are unique locations where the auditor revisited and reassigned water quality scores. With that out of the way, there are a couple of steps we need to take to build the right SQL query to get the information we need from the various tables. Let's start with the following:

1. Grab the `location_id` and `true_water_source_score` attributes from the `auditor_report` entity.
2. **JOIN** the `visits` entity to the auditor_report on the shared `location_id` attribute to get access to the `record_id` attribute.
3. **JOIN** the `water_quality` entity on the `record_id` attribute to retrieve the corresponding `subjective_quality_scores` attribute in the `water_quality` entity.
4. Clean the Resulting table by removing unnecessary redundant columns and renaming the score columns as `surveyor_score` and `auditor_score` respectively
5. Check if the `surveyor_score` and `auditor_score` are different using a `WHERE` clause in the query

In [5]:
%%sql
SELECT 
    a.location_id AS auditor_location,
    a.true_water_source_score AS auditor_score,
    q.subjective_quality_score AS surveyor_score,
    v.record_id
FROM 
    auditor_report AS a
JOIN
    visits AS v
    ON a.location_id = v.location_id 
JOIN
    water_quality AS q
    ON v.record_id = q.record_id
WHERE 
    a.true_water_source_score = q.subjective_quality_score;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
2505 rows affected.


[('SoRu34980', 1, 1, 5185),
 ('AkRu08112', 3, 3, 59367),
 ('AkLu02044', 0, 0, 37379),
 ('AkHa00421', 3, 3, 51627),
 ('SoRu35221', 0, 0, 28758),
 ('HaAm16170', 1, 1, 31048),
 ('AkRu04812', 3, 3, 1513),
 ('AkRu08304', 3, 3, 1218),
 ('AkRu05107', 2, 2, 8322),
 ('HaDe16541', 2, 2, 13070),
 ('KiRu30351', 9, 9, 54220),
 ('AmAm09813', 4, 4, 191),
 ('SoRu37005', 0, 0, 51172),
 ('KiRu26067', 3, 3, 21010),
 ('KiRu26067', 3, 3, 21098),
 ('KiRu26067', 3, 3, 21162),
 ('KiRu26067', 3, 3, 21254),
 ('KiRu26067', 3, 3, 21281),
 ('KiRu26067', 3, 3, 21315),
 ('KiRu26067', 3, 3, 21403),
 ('KiRu26067', 3, 3, 21476),
 ('SoRu37100', 0, 0, 40276),
 ('AmDa11611', 1, 1, 38071),
 ('AkRu07801', 3, 3, 25278),
 ('AkRu07801', 3, 3, 25371),
 ('AkRu07801', 3, 3, 25405),
 ('AkRu07801', 3, 3, 25432),
 ('AkRu07801', 3, 3, 25461),
 ('AkRu07801', 3, 3, 25520),
 ('AkRu07801', 3, 3, 25667),
 ('AkRu07801', 3, 3, 25691),
 ('KiZu31015', 3, 3, 47319),
 ('KiIs24026', 2, 2, 2942),
 ('AkRu08480', 2, 2, 47118),
 ('AmBe11043', 3, 3, 53502),
 ('AmBe11043', 3, 3, 53545),
 ('AmBe11043', 3, 3, 53616),
 ('AmBe11043', 3, 3, 53737),
 ('AmBe11043', 3, 3, 53795),
 ('AmBe11043', 3, 3, 53818),
 ('AmBe11043', 3, 3, 53916),
 ('AmBe11043', 3, 3, 53939),
 ('AmRu13718', 9, 9, 31092),
 ('AmRu15209', 3, 3, 42546),
 ('AmRu15209', 3, 3, 42604),
 ('AmRu15209', 3, 3, 42632),
 ('AmRu15209', 3, 3, 42711),
 ('AmRu15209', 3, 3, 42742),
 ('AmRu15209', 3, 3, 42834),
 ('AmRu15209', 3, 3, 42918),
 ('AmRu15209', 3, 3, 43057),
 ('SoRu37428', 3, 3, 25748),
 ('SoRu37428', 3, 3, 25798),
 ('SoRu37428', 3, 3, 25854),
 ('SoRu37428', 3, 3, 25860),
 ('SoRu37428', 3, 3, 25888),
 ('SoRu37428', 3, 3, 25988),
 ('SoRu37428', 3, 3, 26070),
 ('SoRu37428', 3, 3, 26199),
 ('SoRu34787', 2, 2, 30989),
 ('AkRu03204', 3, 3, 36239),
 ('SoMa34137', 3, 3, 4657),
 ('SoRu38874', 3, 3, 17601),
 ('SoRu38874', 3, 3, 17627),
 ('SoRu38874', 3, 3, 17650),
 ('SoRu38874', 3, 3, 17661),
 ('SoRu38874', 3, 3, 17805),
 ('SoRu38874', 3, 3, 17810),
 ('SoRu38874', 3, 3, 17897),
 ('SoRu38874', 3, 3, 17987),
 ('KiMr24320', 2, 2, 9219),
 ('AkRu06460', 2, 2, 42913),
 ('KiRu28673', 1, 1, 57339),
 ('AkHa00193', 9, 9, 25090),
 ('AmAs10732', 0, 0, 53977),
 ('KiRu27585', 3, 3, 2193),
 ('AkRu03541', 2, 2, 23766),
 ('AkRu04448', 3, 3, 55253),
 ('HaRu20155', 0, 0, 56131),
 ('AmAm09990', 5, 5, 3820),
 ('AmAm09990', 5, 5, 3830),
 ('AmAm09990', 5, 5, 3923),
 ('AmAm09990', 5, 5, 4034),
 ('AmAm09990', 5, 5, 4072),
 ('AmAm09990', 5, 5, 4150),
 ('AmAm09990', 5, 5, 4176),
 ('AmAm09990', 5, 5, 4263),
 ('KiZu31024', 0, 0, 53881),
 ('SoIl32392', 3, 3, 2825),
 ('SoIl32392', 3, 3, 2902),
 ('SoIl32392', 3, 3, 2973),
 ('SoIl32392', 3, 3, 3118),
 ('SoIl32392', 3, 3, 3181),
 ('SoIl32392', 3, 3, 3201),
 ('SoIl32392', 3, 3, 3216),
 ('SoIl32392', 3, 3, 3232),
 ('SoRu38699', 3, 3, 13821),
 ('KiAm22079', 9, 9, 50732),
 ('KiRu26316', 3, 3, 27240),
 ('KiRu26316', 3, 3, 27307),
 ('KiRu26316', 3, 3, 27418),
 ('KiRu26316', 3, 3, 27482),
 ('KiRu26316', 3, 3, 27569),
 ('KiRu26316', 3, 3, 27645),
 ('KiRu26316', 3, 3, 27670),
 ('KiRu26316', 3, 3, 27769),
 ('AkKi01456', 9, 9, 57050),
 ('AkRu08152', 2, 2, 43992),
 ('KiRu27444', 7, 7, 18928),
 ('HaRu20693', 1, 1, 55091),
 ('AkRu04981', 3, 3, 8803),
 ('KiRu28580', 1, 1, 14177),
 ('KiRu30008', 1, 1, 25914),
 ('SoMa34631', 2, 2, 55851),
 ('SoKo33227', 1, 1, 31236),
 ('AmRu14465', 3, 3, 23398),
 ('AmRu14465', 3, 3, 23526),
 ('AmRu14465', 3, 3, 23541),
 ('AmRu14465', 3, 3, 23577),
 ('AmRu14465', 3, 3, 23666),
 ('AmRu14465', 3, 3, 23748),
 ('AmRu14465', 3, 3, 23794),
 ('AmRu14465', 3, 3, 23848),
 ('AkHa00008', 3, 3, 57788),
 ('HaRu19826', 9, 9, 55626),
 ('AmDa11947', 1, 1, 23857),
 ('AmPw12680', 3, 3, 45472),
 ('AmPw12680', 3, 3, 45588),
 ('AmPw12680', 3, 3, 45653),
 ('AmPw12680', 3, 3, 45752),
 ('AmPw12680', 3, 3, 45778),
 ('AmPw12680', 3, 3, 45798),
 ('AmPw12680', 3, 3, 45809),
 ('AmPw12680', 3, 3, 45878),
 ('HaDe16514', 1, 1, 42469),
 ('KiMr24591', 0, 0, 44892),
 ('KiIs24014', 1, 1, 36889),
 ('AkRu07523', 3, 3, 22404),
 ('AkRu04753', 9, 9, 5595

We get a result of **2505** records, but remember that the auditor's records had **1620** records. This is because in the visits entity, various locations were visited more than once meaning that there are duplicate records in our resulting dataset. To remove the duplicate records we specify that we only want records of locations that were visited once in the `WHERE` clause

In [6]:
%%sql
SELECT 
    a.location_id AS auditor_location,
    a.true_water_source_score AS auditor_score,
    q.subjective_quality_score AS surveyor_score,
    v.record_id
FROM 
    auditor_report AS a
JOIN
    visits AS v
    ON a.location_id = v.location_id 
JOIN
    water_quality AS q
    ON v.record_id = q.record_id
WHERE 
    a.true_water_source_score = q.subjective_quality_score
AND 
    v.visit_count = 1;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1518 rows affected.


[('SoRu34980', 1, 1, 5185),
 ('AkRu08112', 3, 3, 59367),
 ('AkLu02044', 0, 0, 37379),
 ('AkHa00421', 3, 3, 51627),
 ('SoRu35221', 0, 0, 28758),
 ('HaAm16170', 1, 1, 31048),
 ('AkRu04812', 3, 3, 1513),
 ('AkRu08304', 3, 3, 1218),
 ('AkRu05107', 2, 2, 8322),
 ('HaDe16541', 2, 2, 13070),
 ('KiRu30351', 9, 9, 54220),
 ('AmAm09813', 4, 4, 191),
 ('SoRu37005', 0, 0, 51172),
 ('KiRu26067', 3, 3, 21010),
 ('SoRu37100', 0, 0, 40276),
 ('AmDa11611', 1, 1, 38071),
 ('AkRu07801', 3, 3, 25278),
 ('KiZu31015', 3, 3, 47319),
 ('KiIs24026', 2, 2, 2942),
 ('AkRu08480', 2, 2, 47118),
 ('AmBe11043', 3, 3, 53502),
 ('AmRu13718', 9, 9, 31092),
 ('AmRu15209', 3, 3, 42546),
 ('SoRu37428', 3, 3, 25748),
 ('SoRu34787', 2, 2, 30989),
 ('AkRu03204', 3, 3, 36239),
 ('SoMa34137', 3, 3, 4657),
 ('SoRu38874', 3, 3, 17601),
 ('KiMr24320', 2, 2, 9219),
 ('AkRu06460', 2, 2, 42913),
 ('KiRu28673', 1, 1, 57339),
 ('AkHa00193', 9, 9, 25090),
 ('AmAs10732', 0, 0, 53977),
 ('KiRu27585', 3, 3, 2193),
 ('AkRu03541', 2, 2, 23766),
 ('AkRu04448', 3, 3, 55253),
 ('HaRu20155', 0, 0, 56131),
 ('AmAm09990', 5, 5, 3820),
 ('KiZu31024', 0, 0, 53881),
 ('SoIl32392', 3, 3, 2825),
 ('SoRu38699', 3, 3, 13821),
 ('KiAm22079', 9, 9, 50732),
 ('KiRu26316', 3, 3, 27240),
 ('AkKi01456', 9, 9, 57050),
 ('AkRu08152', 2, 2, 43992),
 ('KiRu27444', 7, 7, 18928),
 ('HaRu20693', 1, 1, 55091),
 ('AkRu04981', 3, 3, 8803),
 ('KiRu28580', 1, 1, 14177),
 ('KiRu30008', 1, 1, 25914),
 ('SoMa34631', 2, 2, 55851),
 ('SoKo33227', 1, 1, 31236),
 ('AmRu14465', 3, 3, 23398),
 ('AkHa00008', 3, 3, 57788),
 ('HaRu19826', 9, 9, 55626),
 ('AmDa11947', 1, 1, 23857),
 ('AmPw12680', 3, 3, 45472),
 ('HaDe16514', 1, 1, 42469),
 ('KiMr24591', 0, 0, 44892),
 ('KiIs24014', 1, 1, 36889),
 ('AkRu07523', 3, 3, 22404),
 ('AkRu04753', 9, 9, 55958),
 ('SoRu36774', 2, 2, 35173),
 ('SoRu36239', 2, 2, 43524),
 ('KiRu27776', 3, 3, 23770),
 ('AkRu03164', 3, 3, 42866),
 ('AkRu05809', 2, 2, 8004),
 ('KiRu28632', 5, 5, 32143),
 ('KiRu30389', 4, 4, 17755),
 ('AkKi01375', 3, 3, 40201),
 ('SoRu37530', 0, 0, 55624),
 ('AmBe10959', 4, 4, 23719),
 ('HaRu19828', 2, 2, 25621),
 ('SoIl32999', 2, 2, 32992),
 ('HaRu18571', 2, 2, 56000),
 ('AkRu08666', 6, 6, 56276),
 ('AkRu03306', 7, 7, 29135),
 ('AkRu08936', 3, 3, 132),
 ('HaSe21341', 3, 3, 59324),
 ('SoRu35327', 1, 1, 59821),
 ('AmRu14903', 9, 9, 54574),
 ('AmAb09101', 6, 6, 10635),
 ('KiAm22276', 0, 0, 15269),
 ('HaRu20617', 1, 1, 18740),
 ('HaRu18217', 1, 1, 58386),
 ('KiRu29806', 9, 9, 35488),
 ('KiRu28057', 1, 1, 25822),
 ('AkRu06170', 3, 3, 16253),
 ('AkRu03380', 3, 3, 45718),
 ('KiRu26678', 2, 2, 17842),
 ('AmBe10942', 9, 9, 23182),
 ('KiIs23730', 3, 3, 59182),
 ('KiHa23034', 3, 3, 39514),
 ('KiRu28090', 6, 6, 493),
 ('AmRu13729', 2, 2, 12421),
 ('HaRu18016', 1, 1, 43421),
 ('KiRu27440', 9, 9, 46897),
 ('HaYa21606', 3, 3, 54414),
 ('HaRu17207', 1, 1, 41291),
 ('AmAm09735', 1, 1, 19578),
 ('HaRu20534', 2, 2, 36976),
 ('AmPw12770', 3, 3, 57820),
 ('AkLu02096', 3, 3, 46932),
 ('HaRu20434', 2, 2, 26351),
 ('AkRu02688', 0, 0, 28038),
 ('AmBe11285', 9, 9, 30594),
 ('AmRu13720', 3, 3, 57704),
 ('SoMa33819', 0, 0, 41990),
 ('SoRu38119', 1, 1, 32719),
 ('AmRu12983', 3, 3, 34259),
 ('SoRu35906', 5, 5, 23263),
 ('KiRu27742', 3, 3, 20199),
 ('SoRu35991', 3, 3, 49145),
 ('KiRu28083', 4, 4, 28642),
 ('AmRu15159', 9, 9, 3238),
 ('HaYa21456', 2, 2, 31258),
 ('HaRu18734', 2, 2, 9967),
 ('SoRu35970', 3, 3, 49787),
 ('AmAs10285', 3, 3, 35054),
 ('KiZu31223', 3, 3, 45820),
 ('SoRu38546', 9, 9, 31987),
 ('AmPw12293', 6, 6, 43323),
 ('SoRu39488', 3, 3, 778),
 ('SoRu39408', 5, 5, 9997),
 ('HaRu17750', 2, 2, 55863),
 ('KiRu30604', 0, 0, 38650),
 ('KiZu30905', 9, 9, 38879),
 ('SoRu39374', 9, 9, 12919),
 ('SoRu37406', 2, 2, 36882),
 ('AkRu05212', 9, 9, 41421),
 ('SoIl32294', 0, 0, 58139),
 ('AmBe11313', 0, 0, 12271),
 ('SoRu35426', 3, 3, 12319),
 ('AkHa00073', 3, 3, 13554),
 ('SoRu35899', 9, 9, 10054),
 ('KiRu26664', 2, 2, 30561),
 ('SoRu36580', 3, 3, 15465),
 ('AkLu02432', 2, 2, 56526),
 ('HaRu19287', 5, 5, 

Once we've filtered out the duplicate records, we now have a clear view of **1518** surveyor records that match those of the auditor. We can calculate the approximate percentage of accurate scores recorded by the surveyors per the auditor's report.

In [26]:
print(f"Approximately  {round((1518/1620)*100)} % of Surveyoy's records match those of the Auditor")

Approximately  94 % of Surveyoy's records match those of the Auditor


We now know that a huge percentage of the surveyors scores were accurate, what of the remaining ~6%? Let's have a look at those.

## Investigating Difference Patterns in Auditor & Surveyor Scores

We filter the resulting dataset to display records where the `auditor_score` differs from the `surveyor_score`. In our previous analysis, we did not focus on water quality scores but instead investigated various types of water sources.  

To expand our analysis, let’s also include the `type_of_water_source` attribute from the `water_source` entity by **joining** it to our dataset on the `source_id` attribute. This will help ensure there are no discrepancies related to water source classification.
.

In [7]:
%%sql
SELECT 
    a.location_id AS auditor_report,
    a.true_water_source_score AS auditor_score,
    q.subjective_quality_score AS surveyor_score,
    v.record_id,
    s.type_of_water_source AS surveyor_source

FROM 
    auditor_report AS a
JOIN
    visits AS v
    ON a.location_id = v.location_id 
JOIN
    water_quality AS q
    ON v.record_id = q.record_id
JOIN 
    water_source AS s
    ON s.source_id = v.source_id
WHERE 
    a.true_water_source_score != q.subjective_quality_score
AND 
    v.visit_count = 1;
 

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


[('AkRu05215', 3, 10, 21160, 'well'),
 ('KiRu29290', 3, 10, 7938, 'shared_tap'),
 ('KiHa22748', 9, 10, 43140, 'tap_in_home_broken'),
 ('SoRu37841', 6, 10, 18495, 'shared_tap'),
 ('KiRu27884', 1, 10, 33931, 'well'),
 ('KiZu31170', 9, 10, 17950, 'tap_in_home_broken'),
 ('KiZu31370', 3, 10, 36864, 'shared_tap'),
 ('AkRu06495', 2, 10, 45924, 'well'),
 ('HaRu17528', 1, 10, 30524, 'well'),
 ('SoRu38331', 3, 10, 13192, 'shared_tap'),
 ('AmRu15556', 9, 10, 15461, 'tap_in_home_broken'),
 ('AmAm09607', 9, 10, 55067, 'tap_in_home_broken'),
 ('AkHa00314', 9, 10, 29407, 'tap_in_home_broken'),
 ('KiRu26598', 9, 10, 28874, 'tap_in_home_broken'),
 ('KiIs23853', 2, 10, 55796, 'well'),
 ('AkRu03475', 1, 10, 33630, 'well'),
 ('KiIs24083', 2, 10, 28404, 'well'),
 ('AmAm09631', 9, 10, 14211, 'tap_in_home_broken'),
 ('KiRu30343', 2, 10, 51075, 'well'),
 ('AmRu13668', 9, 10, 10586, 'tap_in_home_broken'),
 ('SoBa31691', 0, 10, 3267, 'river'),
 ('AkRu04116', 3, 10, 53182, 'well'),
 ('AmPw12272', 9, 10, 50239, 'tap_in_home_broken'),
 ('SoRu37524', 3, 10, 59640, 'well'),
 ('AmRu14691', 2, 10, 40443, 'well'),
 ('KiRu29556', 2, 10, 5645, 'well'),
 ('KiZu30847', 9, 10, 57038, 'tap_in_home_broken'),
 ('AkHa00363', 0, 10, 25387, 'river'),
 ('SoBa31761', 2, 10, 3296, 'well'),
 ('SoKo33094', 0, 10, 16159, 'river'),
 ('HaSe21323', 3, 10, 26260, 'well'),
 ('AmDa11466', 1, 10, 17086, 'well'),
 ('HaRu17038', 2, 10, 7867, 'well'),
 ('KiRu27364', 0, 10, 38102, 'river'),
 ('AkRu05880', 7, 10, 12705, 'shared_tap'),
 ('KiRu29696', 2, 10, 251, 'well'),
 ('KiRu27065', 0, 10, 29772, 'river'),
 ('AkRu07426', 3, 10, 50918, 'shared_tap'),
 ('AmAs10124', 6, 10, 3154, 'shared_tap'),
 ('HaRu18142', 2, 10, 32555, 'well'),
 ('AkRu06284', 3, 10, 32008, 'shared_tap'),
 ('KiRu25347', 2, 10, 54560, 'well'),
 ('AmRu14364', 2, 10, 42302, 'well'),
 ('KiZu31100', 3, 10, 31033, 'shared_tap'),
 ('KiRu29537', 2, 10, 15869, 'well'),
 ('AkRu03973', 3, 10, 49033, 'well'),
 ('SoKo33333', 2, 10, 56898, 'well'),
 ('KiRu26688', 2, 10, 55794, 'well'),
 ('SoIl32575', 1, 10, 3116, 'well'),
 ('SoMa34366', 9, 10, 26183, 'tap_in_home_broken'),
 ('AkRu04935', 3, 10, 16410, 'shared_tap'),
 ('HaRu19244', 2, 10, 25725, 'well'),
 ('SoRu36708', 3, 10, 7252, 'shared_tap'),
 ('AkRu04508', 1, 10, 37357, 'well'),
 ('AkRu03848', 3, 10, 43412, 'shared_tap'),
 ('AkLu02500', 3, 10, 24797, 'well'),
 ('AkRu07310', 9, 10, 37457, 'tap_in_home_broken'),
 ('AkKi01267', 9, 10, 58866, 'tap_in_home_broken'),
 ('AmRu14245', 9, 10, 47227, 'tap_in_home_broken'),
 ('KiZu30932', 9, 10, 49082, 'tap_in_home_broken'),
 ('KiAm22092', 9, 10, 8186, 'tap_in_home_broken'),
 ('KiRu29147', 0, 10, 3179, 'river'),
 ('SoRu38045', 9, 10, 26566, 'tap_in_home_broken'),
 ('SoIl32770', 0, 10, 7548, 'river'),
 ('HaDj16824', 1, 10, 52429, 'well'),
 ('KiRu29639', 1, 10, 45008, 'well'),
 ('KiRu29329', 0, 10, 11962, 'river'),
 ('AkHa00311', 9, 10, 21392, 'tap_in_home_broken'),
 ('KiMr24919', 1, 10, 23328, 'well'),
 ('HaRu20005', 1, 10, 23715, 'well'),
 ('AmRu14113', 2, 10, 3841, 'well'),
 ('AkKi01256', 1, 10, 34833, 'well'),
 ('SoRu36101', 3, 10, 3351, 'shared_tap'),
 ('SoRu39544', 0, 10, 17929, 'river'),
 ('AkHa00053', 9, 10, 28869, 'tap_in_home_broken'),
 ('HaSe20888', 9, 10, 57001, 'tap_in_home_broken'),
 ('SoRu38401', 2, 10, 20163, 'well'),
 ('AmRu14223', 1, 10, 23126, 'well'),
 ('AmRu14887', 9, 10, 31028, 'tap_in_home_broken'),
 ('AkRu02789', 2, 10, 49443, 'well'),
 ('HaRu20146', 2, 10, 36818, 'well'),
 ('AkRu05741', 3, 10, 19961, 'well'),
 ('KiMr24359', 2, 10, 18443, 'well'),
 ('AkKi00847', 3, 10, 5889, 'well'),
 ('HaRu19621', 3, 10, 55949, 'shared_tap'),
 ('HaYa21395', 2, 10, 28729, 'well'),
 ('AkRu03245', 2, 10, 3233, 'well'),
 ('HaRu20544', 1, 10, 33833, 'well'),
 ('AmRu14959', 9, 10, 1556, 'tap_in_home_broken'),
 ('KiRu30640', 2, 10, 47474, 'well'),
 ('SoMa34211', 1, 10, 24867, 'well'),
 ('SoRu38535', 0, 10, 44282, 'river'),
 ('SoRu36566', 3, 10, 11097, 'shared_tap'),
 ('AmAs10336', 3, 10, 293, 'shared_tap'),
 ('AmRu15719', 9, 10, 34822, 'tap_

As we can see from the resulting dataset, there is a total of **102** records. We can also see that there are no differences between they types of water sources from the auditor report and the surveyors records being compared, so we are safe to proceed with our investigation on that account. Moving forward, the SQL query we created will definitely get more complex so let's see how we can make our lives easier by using **Common Table Expressions (CTEs)**. To do this, we wrap the query we have in a **CTE** and name the expression as `Incorrect_records`. This allows us to break our query down into simpler & more digestable steps. It also enables us to use the referenced query as if it was a table.
**
NOTE: CT**Es are not actual tables hence, they d**o N**OT store data.

While we're on it, let us add `the empl`oyee entity **and **JOIN it to our resulting dataset on `the assigned_employe`e_id attribute so we can have a first look at the surveyors who recorded inaccurate scores per the auditor's report.

In [8]:
%%sql
# Create a CTE of employees/surveyors responsible for the errornours scores
WITH Incorrect_records AS (
    SELECT
        auditor_report.location_id AS audit_location,
        visits.record_id,
        employee.employee_name,
        auditor_report.true_water_source_score AS auditor_score,
        water_quality.subjective_quality_score AS surveyor_score
    FROM
        md_water_services.auditor_report
    JOIN
        md_water_services.visits
        ON visits.location_id = auditor_report.location_id
    JOIN
        md_water_services.water_quality
        ON visits.record_id = water_quality.record_id
    JOIN
        md_water_services.employee
        ON visits.assigned_employee_id = employee.assigned_employee_id
    WHERE 
        visits.visit_count = 1 
        AND auditor_report.true_water_source_score != water_quality.subjective_quality_score
)
SELECT *
FROM Incorrect_records;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


[('AkRu05215', 21160, 'Rudo Imani', 3, 10),
 ('KiRu29290', 7938, 'Bello Azibo', 3, 10),
 ('KiHa22748', 43140, 'Bello Azibo', 9, 10),
 ('SoRu37841', 18495, 'Rudo Imani', 6, 10),
 ('KiRu27884', 33931, 'Bello Azibo', 1, 10),
 ('KiZu31170', 17950, 'Zuriel Matembo', 9, 10),
 ('KiZu31370', 36864, 'Yewande Ebele', 3, 10),
 ('AkRu06495', 45924, 'Bello Azibo', 2, 10),
 ('HaRu17528', 30524, 'Jengo Tumaini', 1, 10),
 ('SoRu38331', 13192, 'Zuriel Matembo', 3, 10),
 ('AmRu15556', 15461, 'Farai Nia', 9, 10),
 ('AmAm09607', 55067, 'Malachi Mavuso', 9, 10),
 ('AkHa00314', 29407, 'Zuriel Matembo', 9, 10),
 ('KiRu26598', 28874, 'Malachi Mavuso', 9, 10),
 ('KiIs23853', 55796, 'Bello Azibo', 2, 10),
 ('AkRu03475', 33630, 'Makena Thabo', 1, 10),
 ('KiIs24083', 28404, 'Lalitha Kaburi', 2, 10),
 ('AmAm09631', 14211, 'Malachi Mavuso', 9, 10),
 ('KiRu30343', 51075, 'Zuriel Matembo', 2, 10),
 ('AmRu13668', 10586, 'Gamba Shani', 9, 10),
 ('SoBa31691', 3267, 'Jengo Tumaini', 0, 10),
 ('AkRu04116', 53182, 'Malachi Mavuso', 3, 10),
 ('AmPw12272', 50239, 'Malachi Mavuso', 9, 10),
 ('SoRu37524', 59640, 'Enitan Zuri', 3, 10),
 ('AmRu14691', 40443, 'Bello Azibo', 2, 10),
 ('KiRu29556', 5645, 'Bello Azibo', 2, 10),
 ('KiZu30847', 57038, 'Bello Azibo', 9, 10),
 ('AkHa00363', 25387, 'Bello Azibo', 0, 10),
 ('SoBa31761', 3296, 'Makena Thabo', 2, 10),
 ('SoKo33094', 16159, 'Lalitha Kaburi', 0, 10),
 ('HaSe21323', 26260, 'Bello Azibo', 3, 10),
 ('AmDa11466', 17086, 'Malachi Mavuso', 1, 10),
 ('HaRu17038', 7867, 'Jengo Tumaini', 2, 10),
 ('KiRu27364', 38102, 'Zuriel Matembo', 0, 10),
 ('AkRu05880', 12705, 'Zuriel Matembo', 7, 10),
 ('KiRu29696', 251, 'Bello Azibo', 2, 10),
 ('KiRu27065', 29772, 'Bello Azibo', 0, 10),
 ('AkRu07426', 50918, 'Rudo Imani', 3, 10),
 ('AmAs10124', 3154, 'Farai Nia', 6, 10),
 ('HaRu18142', 32555, 'Thandiwe Kito', 2, 10),
 ('AkRu06284', 32008, 'Rudo Imani', 3, 10),
 ('KiRu25347', 54560, 'Malachi Mavuso', 2, 10),
 ('AmRu14364', 42302, 'Bello Azibo', 2, 10),
 ('KiZu31100', 31033, 'Bello Azibo', 3, 10),
 ('KiRu29537', 15869, 'Makena Thabo', 2, 10),
 ('AkRu03973', 49033, 'Zuriel Matembo', 3, 10),
 ('SoKo33333', 56898, 'Bello Azibo', 2, 10),
 ('KiRu26688', 55794, 'Rudo Imani', 2, 10),
 ('SoIl32575', 3116, 'Zuriel Matembo', 1, 10),
 ('SoMa34366', 26183, 'Zuriel Matembo', 9, 10),
 ('AkRu04935', 16410, 'Lalitha Kaburi', 3, 10),
 ('HaRu19244', 25725, 'Zuriel Matembo', 2, 10),
 ('SoRu36708', 7252, 'Bello Azibo', 3, 10),
 ('AkRu04508', 37357, 'Bello Azibo', 1, 10),
 ('AkRu03848', 43412, 'Bello Azibo', 3, 10),
 ('AkLu02500', 24797, 'Bello Azibo', 3, 10),
 ('AkRu07310', 37457, 'Lalitha Kaburi', 9, 10),
 ('AkKi01267', 58866, 'Enitan Zuri', 9, 10),
 ('AmRu14245', 47227, 'Zuriel Matembo', 9, 10),
 ('KiZu30932', 49082, 'Malachi Mavuso', 9, 10),
 ('KiAm22092', 8186, 'Bello Azibo', 9, 10),
 ('KiRu29147', 3179, 'Malachi Mavuso', 0, 10),
 ('SoRu38045', 26566, 'Pili Zola', 9, 10),
 ('SoIl32770', 7548, 'Lalitha Kaburi', 0, 10),
 ('HaDj16824', 52429, 'Enitan Zuri', 1, 10),
 ('KiRu29639', 45008, 'Bello Azibo', 1, 10),
 ('KiRu29329', 11962, 'Lalitha Kaburi', 0, 10),
 ('AkHa00311', 21392, 'Usafi Ayo', 9, 10),
 ('KiMr24919', 23328, 'Bello Azibo', 1, 10),
 ('HaRu20005', 23715, 'Malachi Mavuso', 1, 10),
 ('AmRu14113', 3841, 'Yewande Ebele', 2, 10),
 ('AkKi01256', 34833, 'Farai Nia', 1, 10),
 ('SoRu36101', 3351, 'Enitan Zuri', 3, 10),
 ('SoRu39544', 17929, 'Gamba Shani', 0, 10),
 ('AkHa00053', 28869, 'Malachi Mavuso', 9, 10),
 ('HaSe20888', 57001, 'Zuriel Matembo', 9, 10),
 ('SoRu38401', 20163, 'Zuriel Matembo', 2, 10),
 ('AmRu14223', 23126, 'Deka Osumare', 1, 10),
 ('AmRu14887', 31028, 'Bello Azibo', 9, 10),
 ('AkRu02789', 49443, 'Farai Nia', 2, 10),
 ('HaRu20146', 36818, 'Zuriel Matembo', 2, 10),
 ('AkRu05741', 19961, 'Bello Azibo', 3, 10),
 ('KiMr24359', 18443, 'Ona Sefu', 2, 10),
 ('AkKi00847', 5889, 'Zuriel Matembo', 3, 10),
 ('HaRu19621', 55949, 'Yewande Ebele', 3, 10),
 ('HaYa21395', 28729, 'Bello Azibo', 2, 10),
 ('AkRu03245', 3233, 'Malachi Mavuso', 2, 10),
 ('Ha

Our `Incorrect_records` **CTE works**, we notice from the output above that some surveyor names repeat themselves across the records which means that some surveyors messed up more than once. First let's take a look at the unique number of surveyors that recorded inaccurate scores.

In [9]:
%%sql
# Get the number of surveyors responsible for the errornous data
WITH Incorrect_records AS (
    SELECT
        auditor_report.location_id AS audit_location,
        visits.record_id,
        employee.employee_name,
        auditor_report.true_water_source_score AS auditor_score,
        water_quality.subjective_quality_score AS surveyor_score
    FROM
        md_water_services.auditor_report
    JOIN
        md_water_services.visits
        ON visits.location_id = auditor_report.location_id
    JOIN
        md_water_services.water_quality
        ON visits.record_id = water_quality.record_id
    JOIN
        md_water_services.employee
        ON visits.assigned_employee_id = employee.assigned_employee_id
    WHERE 
        visits.visit_count = 1 
        AND auditor_report.true_water_source_score != water_quality.subjective_quality_score
)
SELECT DISTINCT employee_name
FROM Incorrect_records;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
17 rows affected.


[('Rudo Imani',),
 ('Bello Azibo',),
 ('Zuriel Matembo',),
 ('Yewande Ebele',),
 ('Jengo Tumaini',),
 ('Farai Nia',),
 ('Malachi Mavuso',),
 ('Makena Thabo',),
 ('Lalitha Kaburi',),
 ('Gamba Shani',),
 ('Enitan Zuri',),
 ('Thandiwe Kito',),
 ('Pili Zola',),
 ('Usafi Ayo',),
 ('Deka Osumare',),
 ('Ona Sefu',),
 ('Xola Uzuri',)]

**17** Surveyors have errornous records from the output above. We take our analysis further by aggregating the number of times each surveyor recorded inaccurate scores using the `COUNT` function and grouping the results by `employee_name` attribute.

In [10]:
%%sql
# Count the number of times the employees/surveyors responsible for errornous scores made mistakes
WITH Incorrect_records AS (
    SELECT
        auditor_report.location_id AS audit_location,
        visits.record_id,
        employee.employee_name,
        auditor_report.true_water_source_score AS auditor_score,
        water_quality.subjective_quality_score AS surveyor_score
    FROM
        md_water_services.auditor_report
    JOIN
        md_water_services.visits
        ON visits.location_id = auditor_report.location_id
    JOIN
        md_water_services.water_quality
        ON visits.record_id = water_quality.record_id
    JOIN
        md_water_services.employee
        ON visits.assigned_employee_id = employee.assigned_employee_id
    WHERE 
        visits.visit_count = 1 
        AND auditor_report.true_water_source_score != water_quality.subjective_quality_score
)
SELECT
    employee_name,
    COUNT(employee_name) AS number_of_mistakes
FROM 
    Incorrect_records
GROUP BY
    employee_name
ORDER BY number_of_mistakes DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
17 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7),
 ('Rudo Imani', 5),
 ('Farai Nia', 4),
 ('Enitan Zuri', 4),
 ('Yewande Ebele', 3),
 ('Jengo Tumaini', 3),
 ('Makena Thabo', 3),
 ('Gamba Shani', 3),
 ('Thandiwe Kito', 1),
 ('Pili Zola', 1),
 ('Usafi Ayo', 1),
 ('Deka Osumare', 1),
 ('Ona Sefu', 1),
 ('Xola Uzuri', 1)]

We can see that **Bello Azibo** had the highest number of mistakes at **26**, followed by **Malachi Mavuso** at **21**. This might be due to genuine errors or just plain old malice but we don’t know for sure. We need to do some more investigation to see the root cause of the erroneous input from our surveyors. To make our work even simpler, let’s transform the `Incorrect_records` **CTE** into an SQL VIEW and add statements: attribute from the `auditor_report` entity while we’re at it.
 
**NOTE:** Only run the cell below once and then comment out the SQL query since, ideally, the view should be created once.

> **NOTE:** When you restart the notebook and run it again, you will encounter an error since the VIEW already exists in the database by the second run.

In [11]:
# %%sql
# Change the Incorrect_records CTE to a SQL VIEW
# CREATE VIEW Incorrect_records AS (
#     SELECT
#         auditor_report.location_id AS audit_location,
#         visits.record_id,
#         employee.employee_name,
#         auditor_report.true_water_source_score AS auditor_score,
#         water_quality.subjective_quality_score AS surveyor_score,
#         auditor_report.statements AS statements
#     FROM
#         md_water_services.auditor_report
#     JOIN
#         md_water_services.visits
#         ON visits.location_id = auditor_report.location_id
#     JOIN
#         md_water_services.water_quality
#         ON visits.record_id = water_quality.record_id
#     JOIN
#         md_water_services.employee
#         ON visits.assigned_employee_id = employee.assigned_employee_id
#     WHERE 
#         visits.visit_count = 1 
#         AND auditor_report.true_water_source_score != water_quality.subjective_quality_score
# );

Once we've transformed the **CTE** into a **VIEW**, we can test it out to see if our view works as expected.

In [12]:
%sql SELECT * FROM md_water_services.Incorrect_records;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


[('AkRu05215', 21160, 'Rudo Imani', 10, 3, "Villagers admired the official's visit for its respectful interactions, hard work, and genuine concern."),
 ('KiRu29290', 7938, 'Bello Azibo', 10, 3, 'A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.'),
 ('KiHa22748', 43140, 'Bello Azibo', 10, 9, "A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system."),
 ('SoRu37841', 18495, 'Rudo Imani', 10, 6, "The official's respectful and diligent presence was met with heartfelt appreciation, creating a sense of closeness with the villagers."),
 ('KiRu27884', 33931, 'Bello Azibo', 10, 1, "A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community."),
 ('KiZu31170', 17950, 'Zuriel Matembo', 10, 9, 'A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","'),
 ('KiZu31370', 36864, 'Yewande Ebele', 10, 3, "With a keen understanding of urban challenges, the official's visit left a lasting impression of respect and commitment."),
 ('AkRu06495', 45924, 'Bello Azibo', 10, 2, 'A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","'),
 ('HaRu17528', 30524, 'Jengo Tumaini', 10, 1, 'With humility and diligence, the official formed bonds with the villagers that felt like genuine family connections.'),
 ('SoRu38331', 13192, 'Zuriel Matembo', 10, 3, 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('AmRu15556', 15461, 'Farai Nia', 10, 9, "The official's presence left a positive impact on the community, admired for their genuine concern and efforts to improve urban living."),
 ('AmAm09607', 55067, 'Malachi Mavuso', 10, 9, 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('AkHa00314', 29407, 'Zuriel Matembo', 10, 9, "A street vendor's sales suffer from time spent waiting, her concern for the water's quality affecting her products."),
 ('KiRu26598', 28874, 'Malachi Mavuso', 10, 9, "A teenager's dreams are tempered by reality, her future threatened by the corrupt practices she sees around her."),
 ('KiIs23853', 55796, 'Bello Azibo', 10, 2, "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception."),
 ('AkRu03475', 33630, 'Makena Thabo', 10, 1, "Villagers spoke highly of the official's humility and dedication, underscoring their respectful approach and familial connection."),
 ('KiIs24083', 28404, 'Lalitha Kaburi', 10, 2, "A fisherman's anger boils beneath the surface, his livelihood threatened by the corrupt who tamper with water."),
 ('AmAm09631', 14211, 'Malachi Mavuso', 10, 9, 'An elderly man lamented the endless queue, recalling times when water was abundant and clean.","'),
 ('KiRu30343', 51075, 'Zuriel Matembo', 10, 2, "A university student's studies suffer from hours in line, dreams clouded by the daily reality of dirty water."),
 ('AmRu13668', 10586, 'Gamba Shani', 10, 9, "The official's visit left a lasting impact, characterized by their respectful approach and dedication to uplifting the community."),
 ('SoBa31691', 3267, 'Jengo Tumaini', 10, 0, "Villagers were touched by the official's sincere concern and approachable demeanor, creating an atmosphere of unity and understanding."),
 ('AkRu04116', 53182, 'Malachi Mavuso', 10, 3, "A student's ambition is fueled by the daily water struggle, his determination to overcome evident in the queue."),
 ('AmPw12272', 50239, 'Malachi Mavuso', 10, 9, "A subsistence farmer's worry for his crops extends to the queue, the struggle for water a shared community burden."),
 ('SoRu37

Our **VIEW** works perfectly. Let's go ahead and transform the query to calculate the number of errors made by each employee into a **CTE** called `error_count` and use it to compute how many mistakes were made by the **17** surveyors on average using the AVG function.

In [18]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM 
        md_water_services.Incorrect_records 
    GROUP BY
        employee_name
    ORDER BY 
        number_of_mistakes
)
SELECT
    AVG(number_of_mistakes)
FROM
    error_count;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('6.0000'),)]

From the output above, we see that on average, **6** mistakes were made. Let's find out employees that made significant amount of errors by filtering out employees that made more than the average mistakes. we can use a subquery for this one in the `WHERE` clause.

In [21]:
%%sql
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM 
        md_water_services.Incorrect_records 
    GROUP BY
        employee_name
    ORDER BY 
        number_of_mistakes
)
SELECT
    employee_name,
    number_of_mistakes
FROM
    error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
ORDER BY 
    number_of_mistakes DESC
;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Bello Azibo', 26),
 ('Malachi Mavuso', 21),
 ('Zuriel Matembo', 17),
 ('Lalitha Kaburi', 7)]

Now we are only left with **4** surveyors with above average number of mistakes. Since we have the statements recorded by the auditor during his visits to various sites, we can have a look at what locals had to say about these surveyors. First let's wrap the query to get the above average no. of mistakes employees into a **CTE** called `suspect_list` and then use it in a subquery in the `WHERE` clause as demonstrated in the code cell below.

In [29]:
%%sql
# Retrieve statements about the above suspect list
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM
        md_water_services.Incorrect_records
    GROUP BY
        employee_name
    ORDER BY number_of_mistakes DESC
),
suspect_list AS (
    SELECT
        employee_name,
        number_of_mistakes
    FROM
        error_count
    WHERE
        number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
SELECT 
    employee_name,
    location_id,
    statements
FROM
    md_water_services.Incorrect_records
WHERE
    employee_name IN (SELECT employee_name FROM suspect_list);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
71 rows affected.


[('Bello Azibo', 'KiRu29290', 'A young artist sketches the faces in the queue, capturing the weariness of daily hours spent waiting for water.'),
 ('Bello Azibo', 'KiHa22748', "A young girl's hopeful eyes are clouded by mistrust, her innocence tarnished by the corrupt system."),
 ('Bello Azibo', 'KiRu27884', "A traditional healer's empathy turns to bitterness, knowing that corrupt practices harm her community."),
 ('Zuriel Matembo', 'KiZu31170', 'A community leader stood with his people, expressing concern for the water quality and the time lost in queues.","'),
 ('Bello Azibo', 'AkRu06495', 'A healthcare worker in the queue expressed fears about water-borne diseases, her face etched with worry.","'),
 ('Zuriel Matembo', 'SoRu38331', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('Malachi Mavuso', 'AmAm09607', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('Zuriel Matembo', 'AkHa00314', "A street vendor's sales suffer from time spent waiting, her concern for the water's quality affecting her products."),
 ('Malachi Mavuso', 'KiRu26598', "A teenager's dreams are tempered by reality, her future threatened by the corrupt practices she sees around her."),
 ('Bello Azibo', 'KiIs23853', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The mention of cash changing hands further tainted their perception."),
 ('Lalitha Kaburi', 'KiIs24083', "A fisherman's anger boils beneath the surface, his livelihood threatened by the corrupt who tamper with water."),
 ('Malachi Mavuso', 'AmAm09631', 'An elderly man lamented the endless queue, recalling times when water was abundant and clean.","'),
 ('Zuriel Matembo', 'KiRu30343', "A university student's studies suffer from hours in line, dreams clouded by the daily reality of dirty water."),
 ('Malachi Mavuso', 'AkRu04116', "A student's ambition is fueled by the daily water struggle, his determination to overcome evident in the queue."),
 ('Malachi Mavuso', 'AmPw12272', "A subsistence farmer's worry for his crops extends to the queue, the struggle for water a shared community burden."),
 ('Bello Azibo', 'AmRu14691', "An artisan's creative mind is stifled by the wait, thoughts of craft overshadowed by concern for water quality."),
 ('Bello Azibo', 'KiRu29556', "A field worker's endurance is tested by anger, his toil a reminder of the corrupt practices that affect him daily."),
 ('Bello Azibo', 'KiZu30847', "A traveling trader's pause in the village includes a wait for water, his face reflecting the hardship of the road."),
 ('Bello Azibo', 'AkHa00363', "A local mason's strong back is bent by the wait for water, his concerns for quality reflecting a broader challenge."),
 ('Lalitha Kaburi', 'SoKo33094', "A traveling trader's pause in the village includes a wait for water, his face reflecting the hardship of the road."),
 ('Bello Azibo', 'HaSe21323', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('Malachi Mavuso', 'AmDa11466', "A seasonal laborer's exhaustion is compounded by the wait, the water's quality a reminder of the harsh reality."),
 ('Zuriel Matembo', 'KiRu27364', 'A mother of three stood in line for hours, worried about the water\'s color and what it might mean for her children\'s health.","'),
 ('Zuriel Matembo', 'AkRu05880', "Villagers' wary accounts of an official's arrogance and detachment from their concerns raised suspicions. The allusion to cash changing hands deepened their skepticism."),
 ('Bello Azibo', 'KiRu29696', "A grandmother's wisdom and patience in the queue are lessons in resilience, though her concern for water

From our resulting dataset, we can see some statements alluding to surveyor malpractice ranging from a corruption to arrogance. let's zoom in on specific audit locations using the `location_id` in the resulting dataset and see the kinds of statements recorded by the auditor from locals.

In [32]:
%%sql
# Retrieve statements about the above suspect list
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM
        md_water_services.Incorrect_records
    GROUP BY
        employee_name
    ORDER BY number_of_mistakes DESC
),
suspect_list AS (
    SELECT
        employee_name,
        number_of_mistakes
    FROM
        error_count
    WHERE
        number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
SELECT 
    employee_name,
    location_id,
    statements
FROM
    md_water_services.Incorrect_records
WHERE
    employee_name IN (SELECT employee_name FROM suspect_list)
    AND location_id IN ("AkRu04508", "AkRu07310", "KiRu29639", "AmAm09607");

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


[('Malachi Mavuso', 'AmAm09607', 'Villagers spoke of an unsettling encounter with an official who appeared dismissive and detached. The reference to cash transactions added to their growing sense of distrust.'),
 ('Bello Azibo', 'AkRu04508', 'An unsettling atmosphere surrounded the official, as villagers shared their experiences of arrogance and lack of dedication. The mention of cash exchanges only intensified their doubts.'),
 ('Lalitha Kaburi', 'AkRu07310', 'Villagers spoke of their unsettling encounters with an official who seemed indifferent and uninterested, hinting at potential improprieties involving cash exchanges.'),
 ('Bello Azibo', 'KiRu29639', "An unsettling atmosphere prevailed as villagers shared stories of an official's arrogance and perceived corruption. The mention of cash exchanges only intensified their concerns.")]

We've uncovered that surveyor's in our `suspect_list` had a couple of malpractices some taking part in corruption and bribery. Let's confirm to see any other surveyor who also took part in the same but were not included in our `suspect_list`.

In [36]:
%%sql
# Check for any other surveyor who is not in the suspect list with allegations of bribery
WITH error_count AS (
    SELECT
        employee_name,
        COUNT(employee_name) AS number_of_mistakes
    FROM
        md_water_services.Incorrect_records
    GROUP BY
        employee_name
    ORDER BY number_of_mistakes DESC
),
suspect_list AS (
    SELECT
        employee_name,
        number_of_mistakes
    FROM
        error_count
    WHERE
        number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count)
)
SELECT 
    employee_name,
    location_id,
    statements
FROM
    md_water_services.Incorrect_records
WHERE
    statements LIKE "%cash%"
    AND employee_name NOT IN (SELECT employee_name FROM suspect_list);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


[]

## Conclusion

So we can sum up the evidence we have for Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi:

1. They all made more mistakes than their peers on average.
2. They all have incriminating statements made against them, and only them.

Keep in mind, that this is not decisive proof.